doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,294 @@
+
+A "volatile source" is any source that cannot reliabily present the same data
+with the same timestamps across different instantiations.
+
+Examples of volatile sources:
+
+  * Any Kinesis source. Since Kinesis streams have mandatory retention periods
+    (by default 24 hours), a new instantiation of the source will not see
+    any data that has expired.
+  * A PubNub source, since new instantiations always start from the tail of the
+    stream, at least in the current implementation.
+  * A Kafka source with a non-infinite retention policy.
+
+Examples of nonvolatile sources:
+
+  * A file source, when the file is treated as append only.
+  * An S3 source, when the bucket is treated as append only.
+  * An non-upsert Kafka source, when the topic has infinite retention and no
+    compaction policy.
+  * An upsert Kafka source, when the topic has infinite retention but *may*
+    have a key compaction policy.
+
+It's important to note that volatility is a property of how a source is *used*,
+rather than fundamental to a source type. Kinesis is the rare exception. All
+other sources can be either volatile and non-volatile modes, depending on
+how the source is used.
+
+### Discussion
+
+Volatile sources impede Materialize's goal of providing users with correct,
+deterministic results. If you restart Materialize, we want to guarantee that you
+will see exactly the same data at exactly the same timestamps, and there is
+ongoing work to facilitate that. If you've used a volatile source, there is
+simply no possibility that we'll be able to provide that guarantee, because
+the upstream source may have garabage collected the data that you saw last time
+you ran Materialize.
+
+Several forthcoming initiatives depend on this determinism guarantee, however:
+
+  * Exactly-once sinks
+  * Active-active replication
+  * Dataflow transformations that consider source instances to be logically
+    equivalent
+
+My contention is that it is wholly unrealistic to expect our users to only use
+nonvolatile sources with Materialize. The streaming world runs on inconsistency
+and nondurability. To play in this space, we need to interoperate seamlessly
+with volatile sources, and then offer a large, inviting off-ramp into the
+Materialize world of determinism and consistency.
+
+The proposal here, then, is to bake the notion of volatility into the system. We
+start by marking volatile sources as such. Any view that depends on such a
+source becomes volatile. Any sink created from a volatile source or view becomes
+volatile. The optimizer will need to be careful not to treat different
+instantiations of a volatile source or view as logically equivalent. Exactly
+once sinks and other features that only work on nonvolatile sources can look
+for and reject volatile sources.
+
+### Design
+
+To start, I propose that we observe a simple rule:
+
+  * All PubNub, SSE, and Kinesis sources are marked as volatile.
+  * All other sources are not marked as volatile.
```

Comment:
I agree with the proposal as it stands, but at some point we might be able to query the Kafka API and determine if the combination of compaction/retention settings and create source flags mean that the source should be considered volatile.

Reply:
Yeah, it'd be nice to somehow surface to users if their settings don't seem to match their declared volatility. I think the approach of determining this automatically at `CREATE SOURCE` time is going to be hard, though. Folks could always change their retention settings after the fact. It might be nice instead to have a job that does this periodically; like, a cronjob in Materialize that periodically checks to see that the Kafka topic settings upstream are still in a state that match the declared volatility. 

Files and S3 are always going to be hard. Not a whole lot we can do automatically there.

Reply:
I personally feel like it's dangerous to mark some sources as `nonvolatile` that may be marked `volatile` in the future (ex: marking all Kafka sources as `nonvolatile`, although append-only Kafka source with a non-infinite retention policy are, in fact, `volatile`). In my mind, it's worth spending the time to be very specific about `volatile` and `nonvolatile` sources in the V0 of this work. Having source types change under users might undermine our efforts to educate them on the differences, and from a UX perspective a `nonvolatile` to `volatile` switch would be pretty worrying. (I'm less worried about the other direction!)

I think our V0 `nonvolatile` sources should specifically only include what you listed above:
> * An append-only Kafka source, when the topic has infinite retention and no
    compaction policy.
>  * An upsert Kafka source, when the topic has infinite retention but *may*
    have a key compaction policy.
>  * A file source, when the file is treated as append only.

Reply:
> Files and S3 are always going to be hard. Not a whole lot we can do automatically there.

Right, my general position is just that everything that we don't own or validate is sort of in the same equivalence class, some things are just more *obviously* broken than others. Being clear throughout our docs what the effects of folks holding source-dependencies wrong is as important as pointing out that there's no right way to hold a pubnub source.

All the specifics seem good to me, though!

Reply:
> I personally feel like it's dangerous to mark some sources as `nonvolatile` that may be marked `volatile` in the future (ex: marking all Kafka sources as `nonvolatile`, although append-only Kafka source with a non-infinite retention policy are, in fact, `volatile`). In my mind, it's worth spending the time to be very specific about `volatile` and `nonvolatile` sources in the V0 of this work. Having source types change under users might undermine our efforts to educate them on the differences, and from a UX perspective a `nonvolatile` to `volatile` switch would be pretty worrying. (I'm less worried about the other direction!)

So in my mind it's actually going from volatile -> nonvolatile that's dangerous! With a volatile source, the user has basically no responsibility for the data the source produces. But with a nonvolatile source, the user is making some durability promises. If you go from nonvolatile -> volatile, you've removed responsibility from the user, but in the other direction, you're adding responsibility where it didn't previously exist.

> I think our V0 `nonvolatile` sources should specifically only include what you listed above:
>
> * An append-only Kafka source, when the topic has infinite retention and no
compaction policy.
> * An upsert Kafka source, when the topic has infinite retention but may
have a key compaction policy.
> _*_ A file source, when the file is treated as append only.

I'm not sure I entirely follow the proposal! If we want to limit nonvolatile sources to just these, then we'd have to introduce new syntax, because there's no foolproof way for us to tell e.g. whether a file will be treated as append-only. I definitely think that's a worthwhile pursuit, but there are some tricky backwards compatibility concerns to sort out, like: do you mark existing sources volatile or nonvolatile? 

I definitely do want to have a bit of a slower, more deliberate design process where we shake this stuff out! If we change the syntax to e.g. require specifying either `VOLATILE` or `NONVOLATILE`, we'll be forcing users to engage with this idea, and we had better try hard to get it right and get it documented. But at the moment my priority is trying to delicately thread a needle to let us ship PubNub/SSE for the cloud product; marking those source types as volatile without making any permanent syntactic changes feels like a decent balance of tradeoffs; the downside is definitely that we're calling some things nonvolatile that aren't necessarily. (I mean, we're assuming that nonvolatility today, but not quite so explicitly.)

Hmm, I wonder if one easy fix in the meantime is to mark PubNub, Kinesis, and SSE sources as `volatility: true`, but mark all other sources as `volatility: null`. That might be a good way of short-term capturing that we know PubNub, Kinesis, and SSE are known to be volatile, without making assertions about the volatility or non-volatility of the other sources in the system. How does that sound, @JLDLaughlin?

Reply:
> If you go from nonvolatile -> volatile, you've removed responsibility from the user, but in the other direction, you're adding responsibility where it didn't previously exist.

That makes sense to me! I guess I worry that either type of flip is not ideal? Which brings us to...

> Hmm, I wonder if one easy fix in the meantime is to mark PubNub, Kinesis, and SSE sources as volatility: true, but mark all other sources as volatility: null.

I love this! My comment about being specific with the first `VOLATILE`/`NONVOLATILE` sources was a bit more long term (I agree that there's no way to get all the required information now), and I wouldn't want to block the very cool cloud demos. I think this compromise allows us to launch and prevent future whiplash to users.